share/steel/examples/pulse/dice/dpe/DPE.fst

```diff
@@ -0,0 +1,484 @@
+  Destroy the context pointed to by the handle by freeing the
+  arrays in the context (zeroize the UDS, if applicable).
+  NOTE: Current implementation disregards session protocol 
+*)
+```pulse
+fn destroy_context (sid:nat) (ctxt_hndl:nat)
+  requires emp
+  ensures emp
+{
+  let sht = dfst locked_sht;
+  let sht_lk = dsnd locked_sht;
+  W.acquire #(ht_perm sht_sig sht) sht_lk;
+
+  unfold (ht_perm sht_sig sht);
+  with spht. assert (models sht_sig sht spht);
+
+  let opt_locked_cht = PHT.lookup #sht_sig #spht sht sid;
+
+  match opt_locked_cht {
+  Some locked_cht -> {
+    let cht = dfst locked_cht;
+    let cht_lk = dsnd locked_cht;
+    W.acquire #(ht_perm cht_sig cht) cht_lk;
+
+    unfold (ht_perm cht_sig cht);
+    with cpht0. assert (models cht_sig cht cpht0);
+
+    let opt_locked_ctxt = PHT.lookup #cht_sig #cpht0 cht ctxt_hndl;
+    match opt_locked_ctxt {
+    Some locked_ctxt -> {
+      let ctxt = dfst locked_ctxt;
+      let ctxt_lk = dsnd locked_ctxt;
+      W.acquire #(context_perm ctxt) ctxt_lk;
+      match ctxt {
+        Engine_context c -> {
+          rewrite (context_perm ctxt) as (engine_context_perm c);
+          unfold (engine_context_perm c);
+          zeroize_array uds_len c.uds #uds_bytes;
+          disable_uds ();
+          free_array c.uds;
+          fold (ht_perm cht_sig cht);
+          fold (ht_perm sht_sig sht);
+          W.release #(ht_perm cht_sig cht) cht_lk;
+          W.release #(ht_perm sht_sig sht) sht_lk;
+        }
+        L0_context c -> {
+          rewrite (context_perm ctxt) as (l0_context_perm c);
+          unfold (l0_context_perm c);
+          free_array c.cdi;
+          fold (ht_perm cht_sig cht);
+          fold (ht_perm sht_sig sht);
+          W.release #(ht_perm cht_sig cht) cht_lk;
+          W.release #(ht_perm sht_sig sht) sht_lk;
+        }
+        L1_context c -> {
+          rewrite (context_perm ctxt) as (l1_context_perm c);
+          unfold (l1_context_perm c);
+          free_array c.deviceID_priv;
+          free_array c.deviceID_pub;
+          free_array c.aliasKey_priv;
+          free_array c.aliasKey_pub;
+          free_array c.aliasKeyCRT;
+          free_array c.deviceIDCSR;
+          fold (ht_perm cht_sig cht);
+          fold (ht_perm sht_sig sht);
+          W.release #(ht_perm cht_sig cht) cht_lk;
+          W.release #(ht_perm sht_sig sht) sht_lk;
+      }}}
+    None -> {
+      // ERROR - bad context handle
+      fold (ht_perm cht_sig cht);
+      fold (ht_perm sht_sig sht);
+      W.release #(ht_perm cht_sig cht) cht_lk;
+      W.release #(ht_perm sht_sig sht) sht_lk;
+    }}}
+  None -> {
+    // ERROR - bad session ID
+    fold (ht_perm sht_sig sht);
+    W.release #(ht_perm sht_sig sht) sht_lk;
+  }}
+}
+```
+
+```pulse
+fn destroy_cht (ht:ht_t cht_sig)
+  requires ht_perm cht_sig ht
+  ensures emp
+{
+  let mut off = 0sz;
+
+  unfold (ht_perm cht_sig ht);
+  with pht. assert (models cht_sig ht pht);
+  unfold (models cht_sig ht pht);
```

Comment:
You should be able to do this in a single step , e.g.,
```fstar
with pht. unfold (models cht_sig ht pht)
```

Reply:
something strange happening here... the `with ... assert ...` syntax binds pht to the expected type, but `with ... unfold ...` binds pht to unit:

`Expected expression of type "LinearScanHashTable.pht_t (*?u636*) _"; got expression "pht" of type "Prims.unit"`